photon-ml/src/main/scala/com/linkedin/photon/ml/data/AvroDataReader.scala

```diff
@@ -0,0 +1,307 @@
+   *   This configuration merges the "profileFeatures" and "titleFeatures" columns into a single column named
+   *   "userFeatures".
+   * @return the transformed DataFrame
+   */
+  protected def mergeColumns(
+      df: DataFrame,
+      featureColumnMap: Map[String, Set[String]]): DataFrame = {
+
+    val initTempColMappings = Map.empty[String, String]
+
+    // Merge the columns into new temporary columns. We need to create intermediate temporary columns to correctly
+    // handle cases where destination columns are named the same as source columns. Replacing these in-situ could cause
+    // unexpected results when merging subsequent columns.
+    val (merged, tempColMappings) = featureColumnMap.foldLeft((df, initTempColMappings)) {
+      case ((merged, tempColMappings), (destColumn, sourceColumns)) =>
+        val sourceCols = sourceColumns.map(col(_)).toSeq
+        val newCol = mergeFeatures(array(sourceCols:_*))
+        val tempColName = tempColumnName(destColumn)
+
+        (merged.withColumn(tempColName, newCol),
+          tempColMappings ++ Map(destColumn -> tempColName))
+    }
+
+    // Remove the source feature columns
+    val scrubbed = dropColumns(merged, featureColumnMap.values.flatten.toSeq.distinct)
+
+    // Rename the temporary columns
+    tempColMappings.foldLeft(scrubbed) { case (data, (destColumn, tempColName)) =>
+      data.drop(destColumn)
+        .withColumnRenamed(tempColName, destColumn)
+    }
+  }
+
+  /**
+   * Transforms the feature column consisting of name, term, and value feature tuples into a column of feature vectors
+   *
+   * @param df the source DataFrame
+   * @param columnName the feature column to convert into a vector
+   * @param indexMapLoader the index map loader for this column
+   * @return the transformed DataFrame
+   */
+  protected def vectorize(
+      df: DataFrame,
+      columnName: String,
+      indexMapLoader: IndexMapLoader): DataFrame = {
+
+    val vectorized = df.mapPartitions { rows =>
+      val featureMap = indexMapLoader.indexMapForRDD
+
+      rows.map { row: Row =>
+        val columnIndex = row.fieldIndex(columnName)
+        val features = row.getSeq[Row](columnIndex).flatMap {
+          case Row(featureKey: String, value: Double) =>
+            if (featureMap.contains(featureKey)) {
+              Some(featureMap.getIndex(featureKey) -> value)
+            } else {
+              None
+            }
+        }
+
+        // Add intercept if it exists in the feature map
+        val featuresWithIntercept = if (featureMap.contains(AvroDataReader.intercept._1)) {
+          features :+ (featureMap.getIndex(AvroDataReader.intercept._1) -> AvroDataReader.intercept._2)
+        } else {
+          features
+        }
+
+        val (indices, values) = featuresWithIntercept.distinct.sorted.unzip
+        val vector = new SparseVector(featureMap.featureDimension, indices.toArray, values.toArray)
+
+        // Add the column
+        Row.fromSeq(row.toSeq :+ vector)
```

Comment:
The original row contains columns with information about the data point (e.g. offset, weight, response, (key,value) pairs for the features, etc.). This line adds a new column that contains the feature vector transformed from the (key, value) pairs. So originally the DataFrame looks like this:

| response | weight | offset | features |
| --- | --- | --- | --- |
| 1.0 | 1.0 | 0.0 | [("blah1", "", 1.0), ("blah2", "", 1.0)] |

And after this line, it will look like this:

| response | weight | offset | features | features-tmp-col-12345 |
| --- | --- | --- | --- | --- |
| 1.0 | 1.0 | 0.0 | [("blah1", "", 1.0), ("blah2", "", 2.0)] | \<SparseVector 0=1.0 1=2.0\> |

Backing up a bit. In spark.ml, the convention is to have a single column in the DataFrame that contains feature vectors, rather than having columns for each feature. I haven't dug into this design choice in detail, but I presume it's for reasons of sparsity. I'm following their pattern here because it seems like a sane choice, and it will allow easily switching between spark.ml and Photon-ml.

Thanks for bringing this up, I'll add it to the documentation in the code.